src/details/ArborX_DetailsHappyTreeFriends.hpp

```diff
@@ -0,0 +1,61 @@
+/****************************************************************************
+ * Copyright (c) 2017-2021 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_DETAILS_HAPPY_TREE_FRIENDS_HPP
+#define ARBORX_DETAILS_HAPPY_TREE_FRIENDS_HPP
+
+#include <Kokkos_Macros.hpp>
+
+#include <type_traits>
+#include <utility> // declval
+
+namespace ArborX
+{
+namespace Details
+{
+struct HappyTreeFriends
+{
+  template <class BVH>
+  static KOKKOS_FUNCTION auto getNodePtr(BVH const &bvh, int i)
+  {
+    return bvh.getNodePtr(i);
+  }
+
+  template <class BVH>
+  static KOKKOS_FUNCTION auto getRoot(BVH const &bvh)
+  {
+    return bvh.getRoot();
+  }
+
+  template <class BVH>
+  using node_const_pointer_t = decltype(getRoot(std::declval<BVH const &>()));
+
+  template <class BVH>
+  using node_t =
+      std::remove_const_t<std::remove_pointer_t<node_const_pointer_t<BVH>>>;
+
+  template <class BVH>
+  static KOKKOS_FUNCTION decltype(auto)
```

Comment:
I would prefer something more explicit like
```suggestion
  static KOKKOS_FUNCTION typename BVH::bounding_volume_type const&
```
or
```suggestion
  static KOKKOS_FUNCTION auto const&
```
for readability.

Reply:
I meant for the return type to be deduced.  I should probably make the return type of `getRoot()` and `getNodePtr()` also be `decltype(auto)` to be consistent and future proof.

Reply:
> I meant for the return type to be deduced.

I understood that when I saw you using `decltype(auto)`. :slightly_smiling_face: 
Do we anticipate that we would want to return anything else than `typename BVH::bounding_volume_type const&`? Even if, it would be easy to change it later in a backward-compatible way.
The only point I am trying to make is that I prefer to see immediately what the function returns (or at least if it returns by value or reference) if possible without chasing through multiple indirections in the library.

I agree that we should be consistent with whatever we do here.

Reply:
I agree with @masterleinad, it would be nice to clearly state it.